libres/lib/enkf/enkf_main.cpp

```diff
@@ -743,6 +737,337 @@ static FILE *enkf_main_log_step_list(const char *log_path,
+        }
+        parameters[std::string(dataset_name)] = row_scaling_list;
+    }
+    matrix_free(A);
+    return parameters;
+}
+
+static void enkf_main_save_row_scaling_parameters(
+    enkf_fs_type *target_fs, ensemble_config_type *ensemble_config,
+    int_vector_type *iens_active_index, int last_step,
+    const local_ministep_type *ministep,
+    std::unordered_map<
+        std::string,
+        std::vector<std::pair<matrix_type *, const row_scaling_type *>>>
+        parameters) {
+
+    for (auto &[dataset_name, dataset] :
+         local_ministep_get_datasets(ministep)) {
+
+        if (!(local_dataset_get_size(dataset)))
+            continue;
+
+        const auto &scaled_keys = local_dataset_scaled_keys(dataset);
+
+        if (scaled_keys.size() == 0)
+            continue;
+        local_obsdata_type *local_obsdata =
+            local_ministep_get_obsdata(ministep);
+
+        std::vector<std::pair<matrix_type *, const row_scaling_type *>>
+            row_scaling_list = parameters[dataset_name];
+        for (int ikw = 0; ikw < scaled_keys.size(); ikw++) {
+            const auto &key = scaled_keys[ikw];
+            const active_list_type *active_list =
+                local_dataset_get_node_active_list(dataset, key.c_str());
+            matrix_type *A = row_scaling_list[ikw].first;
+            for (int iens = 0; iens < int_vector_size(iens_active_index);
+                 iens++) {
+                int column = int_vector_iget(iens_active_index, iens);
+                if (column >= 0) {
+                    deserialize_node(
+                        target_fs, target_fs,
+                        ensemble_config_get_node(ensemble_config, key.c_str()),
+                        iens, 0, 0, column, active_list, A);
+                }
+            }
+        }
+    }
+}
+
+static void enkf_main_analysis_update_no_rowscaling(
+    analysis_module_type *module, const bool_vector_type *ens_mask,
+    const meas_data_type *forecast, obs_data_type *obs_data,
+    rng_type *shared_rng, matrix_type *E,
+    std::unordered_map<std::string, matrix_type *> parameters) {
+    if (parameters.size() == 0)
+        return;
+    const int cpu_threads = 4;
+    thread_pool_type *tp = thread_pool_alloc(cpu_threads, false);
+    int active_ens_size = meas_data_get_active_ens_size(forecast);
+    int active_size = obs_data_get_active_size(obs_data);
+    matrix_type *X = matrix_alloc(active_ens_size, active_ens_size);
+    matrix_type *S = meas_data_allocS(forecast);
+    matrix_type *R = obs_data_allocR(obs_data);
+    matrix_type *dObs = obs_data_allocdObs(obs_data);
+
+    matrix_type *D = NULL;
+    matrix_type *localA = NULL;
+    int_vector_type *iens_active_index =
+        bool_vector_alloc_active_index_list(ens_mask, -1);
+    const bool_vector_type *obs_mask = obs_data_get_active_mask(obs_data);
+
+    assert_matrix_size(X, "X", active_ens_size, active_ens_size);
+    assert_matrix_size(S, "S", active_size, active_ens_size);
```

Comment:
Consider moving this directly under `matrix_type *S = meas_data_allocS`. 
Same for `assert_matrix_size(R, "R", active_size, active_size);`